test/code/providers/config_scanner_test.cpp

```diff
@@ -0,0 +1,669 @@
+    CPPUNIT_TEST( testCnfScan_IncludeDirWithNonCnfFiles_IgnoreNonCnf );
+    CPPUNIT_TEST( testCnfScan_DirectivesFormCycles);
+    CPPUNIT_TEST( testCnfScan_CnfFileMissing_ReturnsFalse);
+    CPPUNIT_TEST( testCnfScan_InvalidDirectives_Ignored );
+
+    CPPUNIT_TEST_SUITE_END();
+
+private:
+
+public:
+    void setUp(void)
+    {
+    }
+
+    void tearDown(void)
+    {
+    }
+
+    void ValidateAuth(std::map< unsigned int, std::string >& expected, MySQL_Authentication& auth)
+    {
+        std::vector<unsigned int> authPorts;
+        auth.GetPortList( authPorts );
+        CPPUNIT_ASSERT_EQUAL(expected.size(), authPorts.size());
+        for (std::vector<unsigned int>::const_iterator it = authPorts.begin(); it != authPorts.end(); ++it)
+        {
+            MySQL_AuthenticationEntry authEntry;
+            auth.GetEntry(*it, authEntry);
+            CPPUNIT_ASSERT_EQUAL(expected[authEntry.port], authEntry.binding);
+        }
+    }
+
+    void CreateConfig(const std::wstring& binding, unsigned int port, bool withMySqld, std::vector< std::wstring >& conf)
+    {
+        std::wstringstream out;
+        conf.push_back(out.str());
+        conf.push_back(L"# A Comment line");
+        conf.push_back(L"[client]");
+        out.str(L"");
+        out << L"port     = " << port ;
+        conf.push_back(out.str());
+        conf.push_back(L"socket		= /var/run/mysqld/mysqld.sock");
+        conf.push_back(L"An irrelevant line without equalsTo");
+        conf.push_back(L"A line = with more than one =  ");
+        out.str(L"");
+        out << L"bind-address   = " << binding;
+        conf.push_back(out.str());
+        conf.push_back(L"");
+        if (withMySqld)
+        {
+            conf.push_back(L"[mysqld]");
+            conf.push_back(L"#");
+            conf.push_back(L"# * Basic Settings");
+            conf.push_back(L"#");
+            conf.push_back(L"user		= mysql");
+            conf.push_back(L"pid-file	= /var/run/mysqld/mysqld.pid");
+            conf.push_back(L"socket		= /var/run/mysqld/mysqld.sock");
+            conf.push_back(L"port = 3306");
+            conf.push_back(L"basedir		= /usr");
+            conf.push_back(L"datadir		= /var/lib/mysql");
+            out.str(L"");
+        }
+        conf.push_back(out.str());
+    }
+
+    void testCnfScan_NoDirectives()
+    {
+        MySQL_Authentication auth( SCXHandle<MySQL_AuthenticationDependencies>(new MySQL_ScannerTestAuthenticationDependencies()));
+        auth.Load();
+        std::map< std::wstring, std::vector< std::wstring > > fileContentsMap;
+        std::map< std::wstring, std::vector< std::wstring > > directoryMap;
+        std::map< unsigned int, std::string > portBindingMap;
+        std::vector < std::wstring > contents;
+
+        std::wstring cnfFileName = L"NoDirectives.cnf";
+        std::wstring bindAddress = L"172.22.13.1";
+        unsigned int port = 3505;
+        CreateConfig(bindAddress, port, true, contents);
```

Comment:
Addressing your comments:

1. These are open source files. I don't think you need to worry about copyright. Besides, the copyright really belongs to MySQL, and MySQL has an MIT license (very lax). Furthermore, we're not "redistributing" as part of our kit, we're just using them for testing. I don't think #1 is compelling. Furthermore, by not checking actual files from distributions, how do we KNOW we're handling them properly?

2. No, that's not the case. Look at how GetLinuxOS deals with it. The Makefile figures out the "base pwd" for our sources, then builds a directory path to them. This, ultimately, is passed down to GetLinuxOS.sh, which uses those vs. the actual system files on the current system. You can do something virtually identical to that in this case, I think (except as a "prefix" since some files may refer to other files by path).

3. Yes, this is true. But this is true anyway; to verify that data files are processed properly, you need to look at them and see what instances they set up. This isn't a reason to not have robust tests with real-life examples.

4. I don't understand this, sorry. You read the files, then you check the resulting auth object for correctness. It's either correct or it isn't correct. If it isn't correct, then you need to figure out why. To me, there's no better way to handle this than with real-life examples. In this way, we KNOW that actual examples that we include truly work (just like we do with GetLinuxOS.sh). Imagine if GetLinuxOS.sh didn't use real life examples - how would we know it was doing the right thing? I think the same is true in this case.

---
In reply to: [66900085](https://github.com/Microsoft/MySQL-Provider/pull/5#discussion_r66900085) [](ancestors = 66900085)